tools/validate_spec.py

```diff
@@ -0,0 +1,314 @@
+import json
+import sys
+from enum import Enum
+from pathlib import Path
+from typing import Dict, Iterator, List, Optional, Sequence, Union
+
+from rflx.error import RecordFluxError
+from rflx.identifier import ID
+from rflx.pyrflx import PyRFLX, TypeValue
+from rflx.pyrflx.typevalue import MessageValue
+
+
+def cli(argv: List[str]) -> Union[int, str]:
+    parser = argparse.ArgumentParser()
+    subparsers = parser.add_subparsers(dest="subcommand")
+    subparsers.required = True
+
+    parser_verify = subparsers.add_parser(
+        "verify", help="check specification against a set of known valid or invalid examples"
+    )
+
+    parser_verify.add_argument(
+        "-d-valid",
+        "--valid-directory",
+        type=Path,
+        help=("absolute path to directory with known valid examples"),
+        default=None,
+    )
+
+    parser_verify.add_argument(
+        "-d-invalid",
+        "--invalid-directory",
+        type=Path,
+        help=("absolute path to directory with known valid examples"),
+        default=None,
+    )
+
+    parser_verify.add_argument(
+        "-m",
+        "--message-identifier",
+        type=str,
+        help=("identifier of the top-level message: <PackageName>.<MessageName>"),
+        required=True,
+    )
+
+    parser_verify.add_argument(
+        "-o",
+        "--json-output",
+        type=Path,
+        help=("identifier of the top-level message: <PackageName>.<MessageName>"),
+        default=None,
+    )
+
+    parser_verify.add_argument(
+        "--abort-on-error",
+        action="store_true",
+        help=("abort with exitcode 1 if a message cannot be verified correctly"),
+    )
+
+    parser_verify.add_argument("-s", "--specification", type=Path, help=("specification file"))
+
+    parser_verify.set_defaults(func=verify)
+
+    args = parser.parse_args(argv[1:])
+    try:
+        args.func(args)
+    except RecordFluxError as e:
+        return f"{e}"
+
+    return 0
+
+
+def verify(args: argparse.Namespace) -> None:
+    validation_main(
+        args.message_identifier,
+        args.specification,
+        args.invalid_directory,
+        args.valid_directory,
+        args.json_output,
+        args.abort_on_error,
+    )
+
+
+def validation_main(
+    root_message_id: str,
+    path_spec: Path,
+    path_invalid: Optional[Path],
+    path_valid: Optional[Path],
+    full_output: Optional[Path],
+    abort_on_error: bool,
+) -> None:
+    # pylint: disable=too-many-locals
+    def get_message_to_validate(path: Path, is_vaild: bool) -> "Message":
+        if not path.is_file():
+            raise TypeError(f"{path} is not a regular file")
+        message_bytes = path.read_bytes()
+        message = Message(message_bytes, is_vaild, path.name)
+        return message
+
+    if path_valid is None and path_invalid is None:
+        raise ValueError("no directory containing valid or invalid messages has been specified")
+
+    identifier = ID(root_message_id.split("::")).parts
```

Comment:
The `ID` class will do the splitting for you. `ID(root_message_id)` is sufficient. You should catch a `RecordFluxError` here and produce a tool-specific user error from it.

Reply:
I think this has been resolved by https://github.com/Componolit/RecordFlux-specifications/pull/13#discussion_r540362764, but the commit has already been squashed.